network/p2p/dns/resolver.go

```diff
@@ -67,6 +65,26 @@ func WithTTL(ttl time.Duration) optFunc {
 	}
 }
 
+// WithResolverSize is an option function for setting the number of resolver workers and queue size.
+func WithResolverSize(numIPAddrLookupWorkers int, numTxtLookupWorkers int, ipAddrLookupQueueSize int, txtLookupQueueSize int) optFunc {
+	return func(resolver *Resolver) {
+		resolver.ipRequests = make(chan *lookupIPRequest, ipAddrLookupQueueSize)
+		resolver.txtRequests = make(chan *lookupTXTRequest, txtLookupQueueSize)
+		cm := component.NewComponentManagerBuilder()
+
+		for i := 0; i < numIPAddrLookupWorkers; i++ {
+			cm.AddWorker(resolver.processIPAddrLookups)
+		}
+
+		for i := 0; i < numTxtLookupWorkers; i++ {
+			cm.AddWorker(resolver.processTxtLookups)
+		}
+
+		resolver.cm = cm.Build()
+		resolver.Component = resolver.cm
+	}
```

Comment:
Please note that an option function should not have any redundant logic other than just setting the member fields. These lines are a repetition of lines 109-115 which makes the code less DRY (Don't Repeat Yourself principle). We can fix this by moving lines 109-115 post applying the option function, so that we make sure that it adds workers with the new parameters. So the `NewResolver` would look like this: 

```
go

    for _, opt := range opts {
		opt(resolver)
	}

	for i := 0; i < numIPAddrLookupWorkers; i++ {
		cm.AddWorker(resolver.processIPAddrLookups)
	}

	for i := 0; i < numTxtLookupWorkers; i++ {
		cm.AddWorker(resolver.processTxtLookups)
	}

	resolver.cm = cm.Build()
	resolver.Component = resolver.cm
```
Moreover, there is a flaw with the current shape of the code, regardless of the values we pass for the number of workers in the option functions, it always adds workers with the default values in addition to the option function. Refactoring the code with the approach above will fix it. Thanks.